File: src/editor/database/characteristics/ManaType.java
package editor.database.characteristics;

import java.awt.Color;
import java.util.Arrays;
import java.util.Collections;
import java.util.EnumMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

/**
 * This enum represents one of the five colors of Magic: The Gathering.
 * 
 * @author Alec Roelke
 */
public enum ManaType
{
	/**
	 * White mana.
	 */
	WHITE("White", 'W', Color.YELLOW.darker()),
	/**
	 * Blue mana.
	 */
	BLUE("Blue", 'U', Color.BLUE),
	/**
	 * Black mana.
	 */
	BLACK("Black", 'B', Color.BLACK),
	/**
	 * Red mana.
	 */
	RED("Red", 'R', Color.RED),
	/**
	 * Green mana.
	 */
	GREEN("Green", 'G', Color.GREEN.darker()),
	/**
	 * Colorless mana.  While this is not a color, it is a type.
	 */
	COLORLESS("Colorless", 'C', null);

	/**
	 * Get the ManaTypes that represent colors, which is all of them except
	 * {@link #COLORLESS}.
	 * 
	 * @return the colors of Magic, which is the list of ManaTypes
	 * minus the ManaTypes that do not represent colors.
	 */
	public static ManaType[] colors()
	{
		return new ManaType[] {WHITE, BLUE, BLACK, RED, GREEN};
	}
	
	/**
	 * Get a ManaType from a character.  Acceptable characters are 'w,' 'u,' 'b,'
	 * 'r,' 'g,' or 'c,' case insensitive.
	 * 
	 * @param color Character to get a color from
	 * @return ManaType that corresponds to the character
	 * @throws IllegalArgumentException if the character does not correspond to a mana type
	 */
	public static ManaType parseManaType(char color) throws IllegalArgumentException
	{
		ManaType type = tryParseManaType(color);
		if (type == null)
			throw new IllegalArgumentException("Illegal color shorthand '" + color + "'");
		return type;
	}
	
	/**
	 * Get a ManaType from a String.  Acceptable values are "white," "w," "blue,"
	 * "u," "black," "b," "red," "r," "green," "g," "colorless," or "c," case
	 * insensitive.
	 * 
	 * @param color string to create an ManaType from
	 * @return the ManaType that corresponds to the String
	 * @throws IllegalArgumentException if the String does not correspond to a mana type
	 */
	public static ManaType parseManaType(String color) throws IllegalArgumentException
	{
		ManaType type = tryParseManaType(color);
		if (type == null)
			throw new IllegalArgumentException("Illegal color string \"" + color + "\"");
		return type;
	}
	
	/**
	 * Sort a list of ManaTypes in color order.  It is recommended to use this rather than
	 * using Java's built-in sorting functions.
	 * 
	 * @param colors List of ManaTypes to sort
	 */
	public static void sort(List<ManaType> colors)
	{
		Map<ManaType, Integer> counts = new EnumMap<ManaType, Integer>(ManaType.class);
		for (ManaType type: colors)
			counts.compute(type, (k, v) -> v == null ? 1 : v + 1);
		
		List<ManaType> unique = Arrays.stream(colors()).filter(counts::containsKey).collect(Collectors.toList());
		switch (unique.size())
		{
		case 2:
			if (unique.get(0).colorOrder(unique.get(1)) > 0)
				Collections.reverse(unique);
			break;
		case 3:
			while (unique.get(0).distanceFrom(unique.get(1)) != unique.get(1).distanceFrom(unique.get(2)))
				Collections.rotate(unique, 1);
			break;
		case 4:
			ManaType missing = Arrays.stream(colors()).filter((m) -> !counts.containsKey(m)).collect(Collectors.toList()).get(0);
			while (missing.distanceFrom(unique.get(0)) != 1)
				Collections.rotate(unique, 1);
			break;
		default:
			// Don't have to do anything if there are 0, 1, or all 5 colors
			break;
		}
		
		colors.clear();
		if (counts.containsKey(COLORLESS))
			colors.addAll(Collections.nCopies(counts.get(COLORLESS), COLORLESS));
		for (ManaType type: unique)
			colors.addAll(Collections.nCopies(counts.get(type), type));
	}
	
	/**
	 * Get a ManaType from a character.  Acceptable characters are 'w,' 'u,' 'b,'
	 * 'r,' 'g,' or 'c,' case insensitive.
	 * 
	 * @param color Character to get a color from
	 * @return ManaType that corresponds to the character, or null if there is none
	 */
	public static ManaType tryParseManaType(char color)
	{
		for (ManaType c: ManaType.values())
			if (Character.toLowerCase(c.shorthand) == Character.toLowerCase(color))
				return c;
		return null;
	}
	
	/**
	 * Get a ManaType from a String.  Acceptable values are "white," "w," "blue,"
	 * "u," "black," "b," "red," "r," "green," "g," "colorless," or "c," case
	 * insensitive.
	 * 
	 * @param color string to create an ManaType from
	 * @return the ManaType that corresponds to the String, or null if there is none
	 */
	public static ManaType tryParseManaType(String color)
	{
		for (ManaType c: ManaType.values())
			if (c.name.equalsIgnoreCase(color) || color.equalsIgnoreCase(String.valueOf(c.shorthand)))
				return c;
		return null;
	}
	
	/**
	 * Color corresponding to this ManaType (should be null for colorless).
	 */
	public final Color color;
	/**
	 * String representation of this ManaType.
	 */
	private final String name;
	/**
	 * Single-character shorthand for this ManaType.
	 */
	private final char shorthand;
	
	/**
	 * Create a new ManaType.
	 * 
	 * @param n String representation of the new ManaType
	 * @param s single-character shorthand representation of the new ManaType
	 * @param c color corresponding to this ManaType
	 */
	ManaType(final String n, final char s, final Color c)
	{
		name = n;
		shorthand = s;
		color = c;
	}
	
	/**
	 * Compare this ManaType to another ManaType according to the order they would appear in
	 * a mana cost.  The ordering is determined according to the direction around the color pie
	 * in which the distance from this ManaType to the other is shortest.
	 * 
	 * @param other ManaType to compare to
	 * @return a negative number if this ManaType should come first, 0 if they are the same,
	 * or a positive number if it should come after.
	 */
	public int colorOrder(ManaType other)
	{
		if (this == COLORLESS && other == COLORLESS)
			return 0;
		else if (this == COLORLESS)
			return -1;
		else if (other == COLORLESS)
			return 1;
		else
		{
			int diff = compareTo(other);
			return Math.abs(diff) <= 2 ? diff : -diff;
		}
	}
	
	/**
	 * Get the shortest distance around the color pie from this ManaType to the other.
	 * 
	 * @param other ManaType to compare to
	 * @return The distance around the color pie from this color to the other ManaType.
	 */
	public int distanceFrom(ManaType other)
	{
		if (this == COLORLESS || other == COLORLESS)
			throw new IllegalArgumentException("Colorless is not a color");
		return (other.ordinal() - ordinal() + colors().length)%colors().length;
	}
	
	/**
	 * Get the one-character shorthand for this ManaType.
	 * 
	 * @return a one-character shorthand for the name of this ManaType.
	 */
	public char shorthand()
	{
		return shorthand;
	}
	
	@Override
	public String toString()
	{
		return name;
	}
}

File: src/main/scala/editor/database/attributes/ManaType.scala
package editor.database.attributes

import java.awt.Color

/**
 * Object containing global data and parsing methods for [[ManaType]]s.
 * @author Alec Roelke
 */
object ManaType {
  /** List of mana types, in order, that have color. */
  val colors = Seq(White, Blue, Black, Red, Green)

  /**
   * Look up the character corresponding to a [[ManaType]]'s shorthand, ignoring case.
   * 
   * @param s character to look up
   * @return the [[ManaType]] whose shorthand matches the character
   */
  @throws[IllegalArgumentException]("if there is no mana type corresponding to the character")
  def valueOf(s: Char) = values.find(_.shorthand == s.toUpper).getOrElse(throw IllegalArgumentException(s"unknown mana type $s"))

  /**
   * Parse a string for a [[ManaType]]. The string should either contain the whole name of the mana type or
   * be just its single-character shorthand, ignoring case.
   * 
   * @param s string to parse
   * @return the [[ManaType]] whose name or shorthand matches the string, or None if there isn't one
   */
  def parse(s: String) = values.find((c) => c.toString.equalsIgnoreCase(s) || c.shorthand.toString.equalsIgnoreCase(s))

  /**
   * Create a sorted version of the input list of [[ManaType]]s based on their order around the color wheel and how they
   * are typically shown on physical cards.  Use this function to sort a list of [[ManaType]]s rather than any built-in sorting
   * functions. The original list is not changed.
   * 
   * @param types list of types to sort
   * @return a new list containing the same types in the same amounts, but sorted
   */
  def sorted(types: Iterable[ManaType]) = {
    val counts = types.groupBy(identity).map{ case (c, cs) => c -> cs.size }
    val sorted = (counts - Colorless).keys.toSeq.sorted match {
      case Seq(a, b) => if (a.colorOrder(b) < 0) Seq(a, b) else Seq(b, a)
      case s if s.size == 3 =>
        var sorted = s
        while (sorted(0).distanceFrom(sorted(1)) != sorted(1).distanceFrom(sorted(2)))
          sorted = sorted.tail :+ sorted.head
        sorted
      case s if s.size == 4 =>
        val missing = (colors.toSet -- s).head
        var sorted = s
        while (missing.distanceFrom(sorted.head) != 1)
          sorted = sorted.tail :+ sorted.head
        sorted
      case s => s
    }
    Seq.fill(counts.getOrElse(Colorless, 0))(Colorless) ++ sorted.flatMap((c) => Seq.fill(counts(c))(c))
  }
}

/**
 * A type of mana in Magic: The Gathering. Colored mana exists in a wheel that creates a relative ordering of the colors,
 * outside of which is colorless mana, that is depicted on the back of all tournament-legal cards. This wheel can be used
 * to sort lists of mana types. The name of each constant is used to create the long-form name of the mana type.
 * 
 * @constructor create a new mana type
 * @param shorthand single-character representation of mana types in text
 * @param color color to use for depicting the type
 * 
 * @author Alec Roelke
 */
enum ManaType(val shorthand: Char, val color: Option[Color]) extends Ordered[ManaType] {
  import ManaType._

  /**
   * Relative order of this mana type compared to the other one in the shortest direction around the color wheel.
   * 
   * @param other mana type to compare with
   * @return 0 if the two types are the same, a negative number if this type should come before the other type, or a positive
   * number if it should come after
   * @note this should not be used for sorting lists, as it can produce different results depending on the order of entries.  Use [[ManaType.sort]] instead.
   */
  def colorOrder(other: ManaType) = {
    if (this == Colorless && other == Colorless) 0
    else if (this == Colorless) -1
    else if (other == Colorless) 1
    else {
      val diff = ordinal - other.ordinal
      if (math.abs(diff) <= 2) diff else -diff
    }
  }

  /** @return the shortest distance around the color wheel between this type and another */
  def distanceFrom(other: ManaType) = if (this == Colorless || other == Colorless) throw IllegalArgumentException("colorless is not a color") else {
    (other.ordinal - ordinal + colors.length) % colors.size
  }

  override def compare(other: ManaType) = ordinal - other.ordinal

  case Colorless extends ManaType('C', None)
  case White extends ManaType('W', Some(Color.YELLOW.darker))
  case Blue  extends ManaType('U', Some(Color.BLUE))
  case Black extends ManaType('B', Some(Color.BLACK))
  case Red   extends ManaType('R', Some(Color.RED))
  case Green extends ManaType('G', Some(Color.GREEN.darker))
}